sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala

```diff
@@ -0,0 +1,152 @@
+    @Nullable val ret: DataType = inType match {
+      case st: StringType if st.collationId != castType.collationId => castType
+      case ArrayType(arrType, nullable) =>
+        castStringType(arrType, castType).map(ArrayType(_, nullable)).orNull
+      case _ => null
+    }
+    Option(ret)
+  }
+
+  /**
+   * Collates input expressions to a single collation.
+   */
+  def collateToSingleType(exprs: Seq[Expression]): Seq[Expression] = {
+    val st = getOutputCollation(exprs)
+
+    exprs.map(e => castStringType(e, st).getOrElse(e))
+  }
+
+  /**
+   * Based on the data types of the input expressions this method determines
+   * a collation type which the output will have. This function accepts Seq of
+   * any expressions, but will only be affected by collated StringTypes or
+   * complex DataTypes with collated StringTypes (e.g. ArrayType)
+   */
+  def getOutputCollation(expr: Seq[Expression]): StringType = {
+    val explicitTypes = expr.filter(_.isInstanceOf[Collate])
+      .map(_.dataType.asInstanceOf[StringType].collationId)
+      .distinct
+
+    explicitTypes.size match {
+      // We have 1 explicit collation
+      case 1 => StringType(explicitTypes.head)
+      // Multiple explicit collations occurred
+      case size if size > 1 =>
+        throw QueryCompilationErrors
+          .explicitCollationMismatchError(
+            explicitTypes.map(t => StringType(t).typeName)
+          )
+      // Only implicit or default collations present
+      case 0 =>
+        val implicitTypes = expr.map(_.dataType)
+          .filter(hasStringType)
+          .map(extractStringType)
+          .filter(dt => dt.collationId != SQLConf.get.defaultStringType.collationId)
+
+        if (hasMultipleImplicits(implicitTypes)) {
+          throw QueryCompilationErrors.implicitCollationMismatchError()
+        }
+        else {
+          implicitTypes.find(dt => !(dt == SQLConf.get.defaultStringType))
+            .getOrElse(SQLConf.get.defaultStringType)
+        }
+    }
+  }
+
+  /**
+   * This check is always preformed when we have no explicit collation. It returns true
+   * if there are more than one implicit collations. Collations are distinguished by their
+   * collationId.
+   * @param dataTypes
+   * @return
+   */
+  private def hasMultipleImplicits(dataTypes: Seq[StringType]): Boolean =
+    dataTypes.map(_.collationId)
+      .filter(dt => !(dt == SQLConf.get.defaultStringType.collationId)).distinct.size > 1
+
+}
+
+/**
+ * This rule is used to collate all existing expressions related to StringType into a single
+ * collation. Arrays are handled using their elementType and should be cast for these expressions.
+ */
+object PreCollationTypeCasts extends CollationTypeCasts {
+  override val transform: PartialFunction[Expression, Expression] = {
+    case e if !e.childrenResolved => e
+    case sc@(_: In
+             | _: InSubquery
+             | _: CreateArray
+             | _: ComplexTypeMergingExpression
```

Comment:
let's narrow down the scope. Not all `ComplexTypeMergingExpression` expressions require all its inputs to be the same type. Please follow existing implicit cast rules and match If, CaseWhen, etc. explicitly.